README.rst
**************************************
PEGS: Peak-set Enrichment of Gene-Sets
**************************************

``PEGS`` (**P**\ eak-set **E**\ nrichment of **G**\ ene-**S**\ ets) is
a Python bioinformatics utility for calculating enrichments of gene
clusters at different genomic distances from peaks.

.. image:: https://readthedocs.org/projects/pegs/badge/?version=latest
   :target: https://pegs.readthedocs.io/

.. image:: https://badge.fury.io/py/pegs.svg
   :target: https://pypi.python.org/pypi/pegs/

.. image:: https://github.com/fls-bioinformatics-core/pegs/workflows/Python%20package/badge.svg
   :target: https://github.com/fls-bioinformatics-core/pegs/actions?query=workflow%3A%22Python+package%22

* Free software: 3-Clause BSD License
* Documentation: https://pegs.readthedocs.io/en/latest/
* Code: https://github.com/fls-bioinformatics-core/pegs

-----------
Quick Start
-----------

It is recommended to install ``PEGS`` via ``pip`` in a virtualenv,
for example::

    % virtualenv venv.pegs
    % source venv.pegs/bin/activate
    % pip install pegs

This will provide the ``pegs`` and ``mk_pegs_intervals`` programs.

Alternatively you can install ``PEGS`` into a Conda environment,
for example::

    % conda create -c bioconda -c conda-forge -n pegs pegs

The simplest use of ``PEGS`` is:

::

    pegs GENE_INTERVALS --peaks PEAKS1 [PEAKS2 ...] --genes CLUSTERS1 [CLUSTERS2 ...]

where ``GENE_INTERVALS`` is a set of reference transcription
start sites (TSSs) for all genes, ``PEAKS1...N`` are one or more
BED files with peak-sets, and ``CLUSTERS1...N`` are one or more
files defining gene clusters.

This will output a PNG heatmap and an XLXS file with the
p-values and gene counts from the enrichment calculations.

Full documentation can be found at:

 * https://pegs.readthedocs.io/en/latest/

---------
Licensing
---------

PEGS is licensed under the 3-Clause BSD License.

---------------------------------------
Example: Mouse glucocorticoidal dataset
---------------------------------------

*Thanks to Louise Hunter for providing this example.*

An example application of ``PEGS`` is the analysis of up- and
down-regulated glucocorticoidal targets from an RNA-seq study
of liver samples from mice treated acutely with dexamethasone
or vehicle (Caratti *et al.* 2018), combined with corresponding
GR ChIP-seq and chromatin accessibility data (DNase I
hypersensitive (DHS) regions) (Grontved *et al.* 2013 and
Sobel *et al.* 2017 respectively), and mouse liver TAD
boundary data (Kim *et al.* 2018).

The files for these datasets are available here:

 * RNA-seq data:
   `1_zt6up.txt <https://raw.githubusercontent.com/fls-bioinformatics-core/pegs/master/examples/glucocorticoid_example/data/clusters/1_zt6up.txt>`_
   and `2_zt6down.txt <https://raw.githubusercontent.com/fls-bioinformatics-core/pegs/master/examples/glucocorticoid_example/data/clusters/2_zt6down.txt>`_
 * ChIP-seq and chromatin accessibility data:
   `Grontved_GR_ChIP.bed <https://raw.githubusercontent.com/fls-bioinformatics-core/pegs/master/examples/glucocorticoid_example/data/peaks/Grontved_GR_ChIP.bed>`_
   and
   `ZT6_DNase.bed <https://raw.githubusercontent.com/fls-bioinformatics-core/pegs/master/examples/glucocorticoid_example/data/peaks/ZT6_DNase.bed>`_
 * TAD boundary data:
   `mESC-TADs_mm10.txt <https://raw.githubusercontent.com/fls-bioinformatics-core/pegs/master/examples/glucocorticoid_example/data/mESC-TADs_mm10.txt>`_

The ``PEGS`` analysis was performed on these data using the
command::

    pegs mm10 \
      --peaks data/peaks/*.bed --genes data/clusters/*.txt \
      -t data/mESC-TADs_mm10.txt \
      -d 1000 5000 10000 50000 100000 200000 500000 1000000 5000000

The resulting output files (XLSX and heatmap) can be found here:

 * `glucocorticoid_heatmap.png <https://raw.githubusercontent.com/fls-bioinformatics-core/pegs/master/examples/glucocorticoid_example/results/glucocorticoid_heatmap.png>`_
 * `glucocorticoid_results.xlsx <https://raw.githubusercontent.com/fls-bioinformatics-core/pegs/master/examples/glucocorticoid_example/results/glucocorticoid_results.xlsx>`_

.. image:: https://raw.githubusercontent.com/fls-bioinformatics-core/pegs/master/examples/glucocorticoid_example/results/glucocorticoid_heatmap.png

The analyses indicate a strong association of dexamethasone
up-regulated genes with dexamethasone-induced GR peaks at
distances up to 500kbp from these peaks, but no evidence of
down-regulated genes - indicating distinct mechanisms of gene
activation and repression by glucocorticoids. At the same time,
there is promoter proximal enrichment for both up-and

**References**

* Caratti, G. *et al.* (2018) REVERBa couples the clock to hepatic
  glucocorticoid action. *J Clin Invest* **128(10)**:4454-4471
* Grontved, L. *et al.* (2013) C/EBP maintains chromatin accessibility
  in liver and facilitates glucocorticoid receptor recruitment to
  steroid response elements. *EMBO J* **32(11)**, 1568-83
* Kim, Y.H. *et al.* (2018) Rev-erbα dynamically modulates chromatin
  looping to control circadian gene transcription. *Science*
  **359(6381)**:1274-1277
* Sobel, J.A. *et al.* (2017) Transcriptional regulatory logic of the
  diurnal cycle in the mouse liver. *PLoS Biol* **15(4)**: e2001069

----------------
Acknowledgements
----------------

Thanks to Louise Hunter for providing the mouse glucocorticoidal
example, and to Kyle Pollina for allowing us to use the ``pegs``
name on the Python Package Index.


CHANGELOG.rst
Version History and Changes
===========================

--------------------------
Version 0.6.6 (2023-12-07)
--------------------------

* Fix continuous integration testing workflow in Github
  Actions to work with Python 3.6.
* Pin required version of ``xlsxwriter`` to 3.1.9, and
  fix associated bug when creating XLSX files with TADS
  using this version.
* Fix documentation generation on ReadTheDocs.
* Add ``CITATION.cff`` file to GitHub repository (adds
  citation information to the repository front page).

--------------------------
Version 0.6.5 (2022-07-01)
--------------------------

* Always use the ``Agg`` ``matplotlib`` backend (to
  avoid problems with systems where ``DISPLAY``
  environment variable is set but no graphical
  display is available.
* Update documentation for ``mk_pegs_intervals`` to
  clarify how to obtain suitable input files.

--------------------------
Version 0.6.4 (2021-11-08)
--------------------------

* Update the citation information following publication
  of supporting paper on F1000Research
  (https://f1000research.com/articles/10-570/v2)

--------------------------
Version 0.6.3 (2021-10-25)
--------------------------

* Patches to improve handling of input files: fix
  problems with sorting filenames, and perform explicit
  checks on peaks and cluster files to ensure that they
  exist and are not actually directories.
* Update the Conda installation instructions in the
  ``README`` and documentation.

--------------------------
Version 0.6.2 (2021-09-28)
--------------------------

* Update the ``README`` and "getting started"
  documentation to include installation from Conda.

--------------------------
Version 0.6.1 (2021-09-14)
--------------------------

* Patches to update the ``README`` and glucocorticoid
  examples to reflect the  new command line format
  introduced in version 0.6.0.

--------------------------
Version 0.6.0 (2021-09-14)
--------------------------

**This version includes updates that are not backwards
compatible with older versions of PEGS**

* Substantial update to how input files and distances
  are specified:
  - New compulsory ``--peaks`` and ``--genes``
    arguments must now be used to specify one or more
    peakset and gene cluster files respectively.
  - Distance intervals must be specified with new
    ``--distances`` option.
* New options added to customise the output heatmaps:
  - ``--x-label`` and ``--y-label`` options allow custom
    axis labels to be specified.
  - Documentation on setting custom colours for heatmap
    has been expanded (along with examples).
  - Heatmap image format can be specified on the command
    line either implicitly (via output file name extension)
    or explicitly (using ``--format`` option). Supports
    PNG, SVG, PDF etc.
* Added information on how to cite PEGS.

--------------------------
Version 0.5.1 (2021-05-12)
--------------------------

* Patches to fix display of ``README`` landing page on
  PyPI and include the changelog in the documentation.

--------------------------
Version 0.5.0 (2021-05-12)
--------------------------

* Added support for Python 3.9.
* Dropped support for Python 2.7.
* Fixed bug with automatic installation of ``bedtools``
  (was failing on some systems).
* Updated versions of dependencies: ``numpy`` (1.19.5),
  ``scipy`` (1.5.4), ``matplotlib`` (3.3.4) and
  ``seaborn`` (0.11.1).
* Update to installation information to indicate that
  ``PEGS`` can be installed directly from PyPI.

--------------------------
Version 0.4.2 (2021-05-12)
--------------------------

* Updates to ``setup.py`` to enable distributions to be
  built for upload to the Python Package Index (PyPI).

--------------------------
Version 0.4.1 (2021-05-10)
--------------------------

* Fixes to the 'quick start' and 'licensing' sections of
  the ``README``
* Add worked example using mouse glucocorticoidal data to
  the ``README`` and the documentation.
* Update requirement for Pillow library to 8.1.1.
* Fix unit tests and add continuous integration testing
  workflow in Github Actions.

--------------------------
Version 0.4.0 (2020-08-18)
--------------------------

* Drop the requirement that cluster files must be named
  ``cluster_*.txt``; all files in the clusters directory
  will now be included in the analysis. The file names
  (without the file extension) will be used as x-axis
  labels in the plots.

--------------------------
Version 0.3.0 (2020-05-13)
--------------------------

* Automatically install a local version of ``bedtools`` if
  one is not found on the user's ``PATH``.

--------------------------
Version 0.2.0 (2020-03-20)
--------------------------

* Use ``bedtools intersect`` instead of ``intersectBed`` to
  calculate overlaps.
* Cosmetic updates to reporting program progress.
* Add initial Sphinx documentation.

--------------------------
Version 0.1.0 (2020-01-30)
--------------------------

* Initial version of PEGS.


pegs/cli.py
#!/usr/bin/env python
#
#     cli.py: CLI for gene cluster enrichment analysis
#     Copyright (C) University of Manchester 2018-2022 Mudassar Iqbal, Peter Briggs
#
from builtins import str
import os
import argparse
import logging
# Deal with matplotlib backend before importing seaborn
# See https://stackoverflow.com/a/50089385/579925
import matplotlib
matplotlib.use('Agg')
import seaborn as sns
import pathlib2
from .pegs import pegs_main
from .intervals import make_gene_interval_file
from .bedtools import fetch_bedtools
from .bedtools import bedtools_version
from .utils import find_exe
from .utils import collect_files
from .utils import sort_files
from . import get_version

# Description
PEGS_DESCRIPTION = "PEGS: Peak-set Enrichment of Gene-Sets"

# Citation
PEGS_CITATION = """
If you use PEGS in your research then please cite:

* Briggs P, Hunter AL, Yang Sh et al.
  PEGS: An efficient tool for gene set enrichment within defined
  sets of genomic intervals [version 2; peer review: 2 approved].
  F1000Research 2021, 10:570
  (https://doi.org/10.12688/f1000research.53926.2)
"""

# Default set of distances for enrichment calculation
DEFAULT_DISTANCES = [5000,25000,50000,100000,150000,200000]

# Built in gene interval files
BUILTIN_GENE_INTERVALS = {
    "hg38": "refGene_hg38_120719_intervals.bed",
    "mm10": "refGene_mm10_120719_intervals.bed",
}

# Types for cubehelix_palette options
CUBEHELIX_PALETTE_TYPES = {
    'n_colors': int,
    'start': float,
    'rot': float,
    'gamma': float,
    'hue': float,
    'dark': float,
    'light': float,
    'reverse': bool,
}

def pegs():
    # Create command line parser
    p = argparse.ArgumentParser(description=PEGS_DESCRIPTION)
    p.add_argument("gene_intervals",
                   metavar="GENE_INTERVALS",
                   help="either name of a built-in set of gene "
                   "intervals (%s), or a BED file with gene interval "
                   "data" %
                   ','.join(["'%s'" % x for x in BUILTIN_GENE_INTERVALS]))
    p.add_argument('--version',action='version',version=get_version())
    p.add_argument("-p","--peaks",
                   metavar="PEAK_SET_FILE",
                   dest="peaks",
                   action="store",
                   required=True,
                   nargs="+",
                   help="one or more input peak set files (BED format)")
    p.add_argument("-g","--genes",
                   metavar="GENE_CLUSTER_FILE",
                   dest="clusters",
                   action="store",
                   required=True,
                   nargs="+",
                   help="one or more input gene cluster files (one gene "
                   "per line)")
    p.add_argument("-t","--tads",metavar="TADS_FILE",
                   dest="tads_file",
                   action="store",
                   help="BED file with topologically associating "
                   "domains (TADs)")
    p.add_argument("-d","--distances",
                   metavar="DISTANCE",
                   dest="distances",
                   action="store",
                   nargs="+",
                   help="specify distance(s) to calculate enrichments "
                   "for (if no distances are specified then the default "
                   "set will be used i.e. %s)" %
                   ' '.join([str(x) for x in DEFAULT_DISTANCES]))
    output_options = p.add_argument_group("Output options")
    output_options.add_argument("--name",metavar="BASENAME",
                                dest="name",
                                action='store',
                                default="pegs",
                                help="basename for output files (default: "
                                "'pegs')")
    output_options.add_argument("-o",metavar="OUTPUT_DIRECTORY",
                                dest="output_directory",
                                action="store",
                                default=None,
                                help="specify directory to write output "
                                "files to (default: write to current "
                                "directory)")
    output_options.add_argument("-m",metavar="HEATMAP",
                                dest="output_heatmap",
                                action="store",
                                default=None,
                                help="destination for output heatmap; "
                                "image format is implicitly determined by "
                                "the file extension (e.g. '.png','.svg' "
                                "etc) unless overridden by the --format "
                                "option (default: 'BASENAME_heatmap.FORMAT')")
    output_options.add_argument("-x",metavar="XLSX",
                                dest="output_xlsx",
                                action="store",
                                default=None,
                                help="destination for output XLSX file "
                                "with the raw enrichment data (default: "
                                "'BASENAME_results.xlsx')")
    heatmap_options = p.add_argument_group("Heatmap options")
    heatmap_options.add_argument("--format",
                                 dest="heatmap_format",
                                 metavar="FORMAT",
                                 action="store",
                                 default = None,
                                 help="explicitly specify the image format "
                                 "for the output heatmap; note that if this "
                                 "option is specified then it will override "
                                 "the format implied by the specified with "
                                 "the -m option (default: 'png')")
    heatmap_options.add_argument("--x-label",
                                 metavar="CLUSTERS_AXIS_LABEL",
                                 dest="clusters_axis_label",
                                 action="store",
                                 default=None,
                                 help="set a custom label for the X "
                                 "(clusters) axis")
    heatmap_options.add_argument("--y-label",
                                 metavar="PEAKSETS_AXIS_LABEL",
                                 dest="peaksets_axis_label",
                                 action="store",
                                 default=None,
                                 help="set a custom label for the Y "
                                 "(peak sets) axis")
    g = heatmap_options.add_mutually_exclusive_group()
    g.add_argument("--color",
                   dest="heatmap_color",
                   metavar="COLOR",
                   action="store",
                   default=None,
                   help="specify a base color to use for the heatmap "
                   "(NB not compatible with --heatmap-palette)")
    g.add_argument("--heatmap-palette",
                   dest="heatmap_palette_options",
                   metavar="OPTION=VALUE",
                   action="store",
                   nargs="+",
                   default = None,
                   help="advanced option to specify custom palette "
                   "settings for the output heatmap (e.g. 'start=0.5', "
                   "'rot=0' etc). Available options are those listed in "
                   "the 'cubehelix_palette' documentation at "
                   "https://seaborn.pydata.org/generated/"
                   "seaborn.cubehelix_palette.html (NB not compatible "
                   "with --color)")
    advanced_options = p.add_argument_group("Advanced options")
    advanced_options.add_argument("-k","--keep-intersection-files",
                                  dest="keep_intersection_files",
                                  action="store_true",
                                  help="keep the intermediate intersection "
                                  "files (useful for debugging)")
    advanced_options.add_argument("--dump-raw-data",
                                  dest="dump_raw_data",
                                  action="store_true",
                                  help="dump the raw data (gene counts and "
                                  "p-values) to TSV files (for debugging)")
    args = p.parse_args()
    # Deal with peak and cluster files
    peaks = sort_files(args.peaks)
    for f in peaks:
       if not os.path.exists(f):
          logging.fatal("Peaks file '%s' doesn't exist" % f)
          return 1
       elif os.path.isdir(f):
          logging.fatal("Peaks file '%s' is a directory (must be a file)"
                        % f)
          return 1
    clusters = sort_files(args.clusters)
    for f in clusters:
       if not os.path.exists(f):
          logging.fatal("Cluster file '%s' doesn't exist" % f)
          return 1
       elif os.path.isdir(f):
          logging.fatal("Cluster file '%s' is a directory (must be a file)"
                        % f)
          return 1
    # Generate list of distances
    if not args.distances:
        # Defaults
        distances = [d for d in DEFAULT_DISTANCES]
    else:
        # Assemble from command line
        distances = list()
        for d in args.distances:
            for x in d.split(','):
                distances.append(int(x))
    distances = sorted(distances)
    # Check if using built-in interval data
    gene_interval_file = args.gene_intervals
    try:
        gene_interval_file = BUILTIN_GENE_INTERVALS[gene_interval_file]
        p = os.path.dirname(__file__)
        while p != os.sep:
            f = os.path.join(p,"pegs-%s" % get_version(),gene_interval_file)
            if os.path.exists(f):
                gene_interval_file = f
                break
            else:
                p = os.path.dirname(p)
    except KeyError:
        # Not found, ignore
        pass
    # Check TADs file is actually a file
    if args.tads_file:
       if not os.path.exists(args.tads_file):
          logging.fatal("TADs file '%s' doesn't exist" % args.tads_file)
          return 1
       elif os.path.isdir(args.tads_file):
          logging.fatal("TADs file '%s' is a directory (must be a file)"
                        % args.tads_file)
          return 1
    # Build colormap for heatmap
    heatmap_cmap = None
    if args.heatmap_color:
        # Construct non-default colormap using the
        # seaborn lightpalette function
        heatmap_cmap = sns.light_palette(color=args.heatmap_color,
                                         as_cmap=True)
    elif args.heatmap_palette_options is not None:
        # Construct non-default colormap using the
        # options supplied by the user
        heatmap_palette_options = {
            'n_colors': 6,
            'start': 0,
            'rot': 0.4,
            'gamma': 1.0,
            'hue': 0.8,
            'light': 0.85,
            'dark': 0.15,
            'reverse': False,
        }
        for o in args.heatmap_palette_options:
            key,value = o.split("=")
            if key not in heatmap_palette_options:
                logging.warning("Unrecognised palette option: '%s'"
                                % key)
            else:
                heatmap_palette_options[key] = \
                        CUBEHELIX_PALETTE_TYPES[key](value)
        heatmap_cmap = sns.cubehelix_palette(as_cmap=True,
                                             **heatmap_palette_options)
    # Report version and authors etc
    print("%s %s" % (PEGS_DESCRIPTION,get_version()))
    print("""
Efficiently calculate enrichments of gene clusters in
multiple genomic intervals data (e.g. ChIP-seq peak-sets)
at different distances

Copyright University of Manchester
Faculty of Biology Medicine and Health
Authors: Mudassar Iqbal, Peter Briggs
""")
    print(PEGS_CITATION)

    print("====PEGS is starting====")

    # Add PEGS 'bin' directory in user's home area to PATH
    # NB this might not exist
    pegs_dir = os.path.join(str(pathlib2.Path.home()),".pegs")
    pegs_bin_dir = os.path.join(pegs_dir,"bin")
    os.environ['PATH'] = "%s%s%s" % (os.environ['PATH'],
                                     os.pathsep,
                                     pegs_bin_dir)

    # Locate bedtools executable
    bedtools_exe = find_exe("bedtools")
    if not bedtools_exe:
        # Not found
        logging.warning("'bedtools' not found")
        # Attempt to get bedtools
        bedtools_exe = fetch_bedtools(install_dir=pegs_bin_dir,
                                      create_install_dir=True)
        if not bedtools_exe:
            logging.fatal("Failed to fetch 'bedtools'")
            return 1
    print("Found %s (%s)\n" % (bedtools_version(bedtools_exe),
                               bedtools_exe))

    # Calculate the enrichments
    pegs_main(genes_file=gene_interval_file,
              distances=distances,
              peaks=peaks,
              clusters=clusters,
              tads_file=args.tads_file,
              name=args.name,
              heatmap=args.output_heatmap,
              xlsx=args.output_xlsx,
              output_directory=args.output_directory,
              keep_intersection_files=
              args.keep_intersection_files,
              clusters_axis_label=args.clusters_axis_label,
              peaksets_axis_label=args.peaksets_axis_label,
              heatmap_cmap=heatmap_cmap,
              heatmap_format=args.heatmap_format,
              dump_raw_data=args.dump_raw_data)

def mk_pegs_intervals():
    # Create command line parser
    p = argparse.ArgumentParser()
    p.add_argument("refgene_file",
                   metavar="REFGENE_FILE",
                   help="refGene annotation data for the genome "
                   "of interest")
    p.add_argument("gene_interval_file",
                   metavar="GENE_INTERVAL_FILE",
                   nargs='?',
                   help="destination for output BED file with "
                   "gene interval data (default: "
                   "'<REFGENE_FILE>_intervals.bed')")
    p.add_argument('--version',action='version',version=get_version())
    args = p.parse_args()
    # Report version
    print("MK_PEGS_INTERVALS %s\n" % get_version())
    # Generate the gene interval file
    make_gene_interval_file(args.refgene_file,
                            args.gene_interval_file)
